### PR TITLE
New ingress <-> PP communication

### PR DESCRIPTION
**Pull request by slinkydeveloper**

### CODE REVIEW

File: crates/worker/src/partition/leadership.rs
Comment (tillrohrmann) [reply crates/worker/src/partition/leadership.rs:99]: What happens with the awaiting rpcs if we become a follower?

File: crates/worker/src/partition/mod.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/mod.rs:588]: Once `self_propose` completes it is not guaranteed that the record was appended to Bifrost.

File: crates/worker/src/partition/mod.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/mod.rs:691]: At this point, it is not guaranteed that the invocation was appended to Bifrost when using the background appender.

File: crates/worker/src/partition/mod.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/mod.rs:759]: Could we also return `NotReady` if the status is neither free nor completed independent of this special casing?

File: crates/worker/src/partition/state_machine/mod.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/state_machine/mod.rs:1728]: Are we still needing the `Ingress` variant? Given that a rpc request is ephemeral, we probably don't need to persist that an ingress is waiting for a response.

File: crates/worker/src/partition/state_machine/mod.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/state_machine/mod.rs:2809]: For what do we need the `target_node`? Wouldn't this be encoded in the `awaiting_rpcs` map in the `LeaderState`?

File: crates/worker/src/partition_processor_manager.rs
Comment (tillrohrmann) [inline crates/worker/src/partition_processor_manager.rs:571]: Nice :-)

**Review by tillrohrmann — COMMENTED** [review]
Thanks a lot for creating this PR @slinkydeveloper. The changes look good to me.

One interesting piece which is missing for an end-to-end review is the way we handle messages that are sent to non-leaders and the retry policy (time base retries vs. event based retries on connection loss). I guess we could already start with it even w/o a fully functioning routing layer (should come soon via https://github.com/restatedev/restate/pull/2149).

I had a few questions around:

* What information do we need to persist with this change? Do we still need to record the `InvocationResponseSink::Ingress`?
* Could we use the message id instead of the `PartitionProcessorRpcRequestId` for correlation?
* Do we need to send the RPC request id information through Bifrost or is there a different way to correlate events with pending rpcs on the PartitionProcessor (e.g. using `InvocationId`)?

One thing which should not properly work atm is the response which send for an `append` request. I think we are sending the response before the record is committed to Bifrost.

File: crates/worker/src/partition/leadership.rs
Comment (slinkydeveloper) [reply crates/worker/src/partition/leadership.rs:704]: yep

File: crates/worker/src/partition/mod.rs
Comment (slinkydeveloper) [reply crates/worker/src/partition/mod.rs:588]: how do i guarantee that? I guess I need to stop using the bifrost background appender or what?

Comment (tillrohrmann) [reply crates/worker/src/partition/mod.rs:588]: You can call `enqueue_with_notification` which gives you a future to await on. However, this should not happen as part of this method since it would block the `PartitionProcessor` from processing other requests.

File: crates/worker/src/ingress_integration.rs
Comment (slinkydeveloper) [reply crates/worker/src/ingress_integration.rs:65]: I think this specific case it's ok, it's very isolated to the ingress. This is also same behavior as before.

**Comment by slinkydeveloper** [comment]
Gonna do a first squash to make the rebase reasonable :roll_eyes:

File: crates/core/src/network/partition_processor_rpc_client.rs
Comment (slinkydeveloper) [reply crates/core/src/network/partition_processor_rpc_client.rs:87]: Fixed a bit the naming in this interface

File: crates/core/src/network/partition_processor_rpc_client.rs
Comment (slinkydeveloper) [reply crates/core/src/network/partition_processor_rpc_client.rs:112]: Yes, as discussed offline the easier solution is to retain the identity of the caller throughout the whole chain.

File: crates/worker/src/partition/mod.rs
Comment (slinkydeveloper) [reply crates/worker/src/partition/mod.rs]: Done

File: crates/worker/src/partition/leadership.rs
Comment (slinkydeveloper) [reply crates/worker/src/partition/leadership.rs:633]: As discussed offline no for now

File: crates/worker/src/partition/mod.rs
Comment (slinkydeveloper) [reply crates/worker/src/partition/mod.rs:728]: I would like to retain this special casing, as back in the handler we use this special case to give back the user a nicer descriptive error. Look at `HandlerError::UnsupportedGetOutput`.

File: crates/core/src/network/partition_processor_rpc_client.rs
Comment (slinkydeveloper) [reply crates/core/src/network/partition_processor_rpc_client.rs:41]: I prefer to avoid using this pattern here, because i would have to leak abstractions back and forth and I don't really like it. I'm ok with cloning, eventually ServiceInvocation IMO should all go behind an arc or something (and already good part of it it's behind an arc). If this becomes a performance issue, i'll deal with it.

**Comment by slinkydeveloper** [comment]
I think this is now ready for another review. As discussed offline, we still have the concept of http request caller identity in the partition processor through this `PartitionProcessorRpcId`. I've also added the basic retry/failure logic of RPCs within `RpcRequestDispatcher`, where we should incrementally add all the "magic" of this interaction (while I would prefer `PartitionProcessorClient` remains a simple facade on the messages and the network client)

File: crates/worker/src/partition/leadership.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/leadership.rs:586]: Should this be renamed into `pending_cleanup_timers_to_schedule`?

File: crates/worker/src/partition/leadership.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/leadership.rs:693]:
```suggestion
                    |pending_cleanup_timers_to_schedule| {
```

File: crates/worker/src/partition/leadership.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/leadership.rs:706]: Nice :-)

File: crates/worker/src/partition/leadership.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/leadership.rs:923]: Could be a `if let` instead of a `take` to cover the case that we already sent an error.

File: crates/worker/src/partition/leadership.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/leadership.rs:939]: Nice simplification with the `SelfProposer`.

File: crates/worker/src/partition/mod.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/mod.rs:639]: I think here is a potential for a race condition. Assume that the invocation is not ready but we enqueue the `AttachInvocation` after an `End` command. In this case, we won't respond with the result if it is not an idempotent invocation.

File: crates/worker/src/partition/mod.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/mod.rs:731]: State why it safe to use unsafe here?

File: crates/worker/src/partition/state_machine/mod.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/state_machine/mod.rs:508]: What if we handled non-idempotent invocations the same way we treat idempotent invocations? Like if an invocation is already present, then we update the `response_sinks` with then new `service_invocation.response_sink` and send a submit notification? This would make submitting invocations idempotently based on their invocation id independent of the idempotency key.

Now, we are already half-way there with this change.

**Review by tillrohrmann — COMMENTED** [review]
Thanks for updating the PR @slinkydeveloper. I think we are almost at the point where we can merge it. The changes look good to me. The one thing I will try to figure out is the ingress retry behavior. I'll first rebase your PR onto the latest main to get started.

File: crates/worker/src/partition/mod.rs
Comment (tillrohrmann) [reply crates/worker/src/partition/mod.rs:639]: I think this is not a real problem so please ignore this comment.

File: crates/worker/src/partition/state_machine/mod.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/state_machine/mod.rs:584]: I have to admit that I have a hard time following the flow of the `on_service_invocation` call and the methods it calls. Therefore, it's hard for me to assert the correctness. One problem that I see is that we send messages at different places (e.g. the submit notification) and based on some return value expect the caller to continue or not. Due to this property, one needs to follow every branch and check that all the required actions happen and when popping the stack on needs to remember based on the return value that certain things need to happen or not. 

One idea that could simplify things a bit is to do the I/O at one place (I guess in the `on_service_invocation` method).

File: crates/worker/src/partition/state_machine/tests/mod.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/state_machine/tests/mod.rs:999]: Should we also check that the send the correct `SubmittedInvocationNotification` with the correctly set `is_new_invocation`?

**Review by tillrohrmann — COMMENTED** [review]
Thanks a lot for your enduring work on this PR @slinkydeveloper. I think it's really good 🚀  There are two things we probably need discuss before merging:

1. Is the ingress variant change to the `Source` proto backward compatible?
2. Right now, it looks to me that `NotLeader` is not safe to retry because we also get it after a leader loses leadership. I think this can cause duplicate messages.

A less important question is whether we still need the `DispatchIngressRequest` or can merge the remaining functionality with the Kafka consumer task.

File: crates/core/src/network/rpc_router.rs
Comment (slinkydeveloper) [reply crates/core/src/network/rpc_router.rs:51]: it didn't through other traits, b/c it was requiring `T: Clone`. This is a pitfall of the derive `Clone` feature in rust, will require all the generics to be clone as well, even though this is not needed sometimes.

File: crates/ingress-dispatcher/src/dispatcher.rs
Comment (slinkydeveloper) [reply crates/ingress-dispatcher/src/dispatcher.rs:75]: Not really used anymore, but i'll cleanup the kafka/dispatcher thing later on.

File: crates/ingress-dispatcher/src/dispatcher.rs
Comment (slinkydeveloper) [reply crates/ingress-dispatcher/src/dispatcher.rs:31]: For now I keep everything as it is in this module, I solve it in a followup.

File: crates/ingress-http/src/lib.rs
Comment (slinkydeveloper) [reply crates/ingress-http/src/lib.rs:54]: True, removed this comment

File: crates/ingress-http/src/lib.rs
Comment (slinkydeveloper) [reply crates/ingress-http/src/lib.rs:70]: Same as above

File: crates/storage-api/proto/dev/restate/storage/v1/domain.proto
Comment (slinkydeveloper) [reply crates/storage-api/proto/dev/restate/storage/v1/domain.proto:95]: As discussed offline yes, i'll make the conversion a soft error and we should be good.

File: crates/worker/src/ingress_integration.rs
Comment (slinkydeveloper) [reply crates/worker/src/ingress_integration.rs:46]: I'm not sure honestly which parameters to pick for the exponential? Surely I wanna have a short max-interval i think.

Comment (slinkydeveloper) [reply crates/worker/src/ingress_integration.rs:46]: Maybe linear is actually fine and we just need some jitter?

File: crates/worker/src/partition/leadership.rs
Comment (slinkydeveloper) [reply crates/worker/src/partition/leadership.rs:533]: Done, can you perhaps double check the cases again?

File: crates/worker/src/partition/leadership.rs
Comment (slinkydeveloper) [reply crates/worker/src/partition/leadership.rs:864]: Then I prefer to not do it for now.

File: crates/worker/src/partition/state_machine/mod.rs
Comment (slinkydeveloper) [reply crates/worker/src/partition/state_machine/mod.rs:535]: I'm afraid this will trigger also if the user goes first back and then forward with versions, now that I made the Source::Ingress parsing return a random id, in case one is not present.

File: crates/worker/src/partition/state_machine/mod.rs
Comment (slinkydeveloper) [reply crates/worker/src/partition/state_machine/mod.rs:584]: For now I'm not gonna take any action regarding this, I plan in future to reorganize this code a bit.

File: crates/worker/src/partition/state_machine/mod.rs
Comment (slinkydeveloper) [reply crates/worker/src/partition/state_machine/mod.rs:535]: So i'm actually inclined to just remove it...

Comment (tillrohrmann) [reply crates/worker/src/partition/state_machine/mod.rs:535]: I think removing it should be fine.

File: crates/core/src/network/partition_processor_rpc_client.rs
Comment (tillrohrmann) [inline crates/core/src/network/partition_processor_rpc_client.rs:102]: So I can't attach to a running non-idempotent invocation?

File: crates/worker/src/partition/leadership.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/leadership.rs:535]: `fail_with_lost_leadership`?

File: crates/worker/src/partition/leadership.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/leadership.rs:899]: `fail_with_lost_leadership`

File: crates/worker/src/partition/state_machine/tests/mod.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/state_machine/tests/mod.rs:1028]: The comment seems wrong now.

File: crates/worker/src/partition/state_machine/tests/mod.rs
Comment (tillrohrmann) [inline crates/worker/src/partition/state_machine/tests/mod.rs]: Shouldn't this return `true` if the `Invoke` command was a duplicate of the original request?

**Review by tillrohrmann — APPROVED** [review]
🥇 LGTM. +1 for merging.

File: crates/core/src/network/partition_processor_rpc_client.rs
Comment (slinkydeveloper) [reply crates/core/src/network/partition_processor_rpc_client.rs:102]: No, that's the current semantics.

File: crates/worker/src/partition/state_machine/tests/mod.rs
Comment (slinkydeveloper) [reply crates/worker/src/partition/state_machine/tests/mod.rs]: Yep, fixed this

File: crates/core/src/network/partition_processor_rpc_client.rs
Comment (tillrohrmann) [reply crates/core/src/network/partition_processor_rpc_client.rs:102]: Is this contract documented somewhere? In our docs I could only find https://docs.restate.dev/develop/java/clients/#retrieve-result-of-invocations-and-workflows which does not do any distinction between idempotent and non-idempotent invocations. Why wouldn't it be possible to attach to a non-idempotent invocation that I initially submitted via send? Isn't it the same as if I submit an idempotent invocation with retention time of 0?

**Pull request merged**
